File: Ch03. Basic Data Structures/Java code/LinkedListsSingle/LinkedListSingle.java
public class LinkedListSingle {

    private Node head, tail;

    public LinkedListSingle(){
        this.head = this.tail = null;
    }
    public boolean isEmpty(){
        return this.head == null;
    }
    public int size(){
        int counter = 0;
        for(Node tmp = this.head; tmp != null; tmp = tmp.next)
            counter++;
        return counter;
    }
    public void addToHead(int i){
        this.head = new Node(i, this.head);
        if(this.tail == null)
            this.tail = this.head;
    }
    public void addToTail(int i){
        if(this.isEmpty())
            this.head =this.tail = new Node(i);
        else {
            this.tail.next = new Node(i);
            this.tail = this.tail.next;
        }
    }
    public Integer deleteFromHead(){
        if(this.isEmpty())
            return null;
        else {
            Integer i = this.head.info;
            if(this.head == this.tail)
                this.head = this.tail = null;
            else {
                this.head = this.head.next;
            }
            return i;
        }
    }
    public Integer deleteFromTail(){
        if(this.isEmpty())
            return null;
        else{
            Integer i = this.tail.info;
            if(this.head == this.tail)
                this.head = this.tail = null;
            else {
                Node tmp;
                for(tmp = this.head; tmp.next != this.tail; tmp = tmp.next);
                tmp.next = null;
                this.tail = tmp;
            }
            return i;
        }
    }
    public void printAll(){
        for(Node tmp = this.head; tmp != null; tmp = tmp.next)
            System.out.print(tmp.info + ", ");
    }
    public void deleteNodesWithValue(int value){
        if(this.isEmpty())
            return;
        else if(this.head != this.tail){
            Node currentNode = this.head;
            while(currentNode.next != null){
                if(currentNode.next.info == value){
                    if(currentNode.next.next == null)
                        currentNode.next = null;
                    else 
                        currentNode.next = currentNode.next.next;
                }
                else
                    currentNode = currentNode.next;
            }
            this.tail = currentNode;
        }  
        if(this.head.info == value){
            if(this.head == this.tail)
                this.head = this.tail = null;
            else 
                this.head = this.head.next;
        } 
    }
    public void sort(){
        for(Node outer = this.head; outer != null; outer = outer.next){
            boolean swapped = true;
            do{
                swapped = false;
                for(Node inner = this.head; inner.next != null; inner = inner.next){
                    if(inner.info > inner.next.info){
                        int tmp = inner.info;
                        inner.info = inner.next.info;
                        inner.next.info = tmp;
                        swapped = true;
                    }
                }
            }
            while(swapped == true);
        }
    }
}

File: Ch03. Basic Data Structures/SigleLinkedList/LinkedListSingle.py
from Node import Node

class LinkedListSingle:
    def __init__(self):
        self.head = None
        self.tail = None

    def isEmpty(self):
        return self.head == None

    def size(self):
        counter = 0
        tmp = self.head
        while tmp != None:
            counter += 1
            tmp = tmp.next
        return counter

    def printAll(self):
        result = ""
        tmp = self.head
        while tmp != None:
            result += str(tmp.payload)
            tmp = tmp.next
        print(result)

    def addToHead(self, payload):
        self.head = Node(payload, self.head)
        if self.tail == None:
            self.tail = self.head
    
    def addToTail(self, payload):
        if self.isEmpty():
            self.head = self.tail = Node(payload)
        else:
            self.tail.next = Node(payload)
            self.tail = self.tail.next

    def deleteFromHead(self):
        if self.isEmpty():
            return
        else:
            if self.head == self.tail:
                self.head = self.tail = None
            else:
                self.head =self.head.next

    def deleteFromTail(self):
        if self.isEmpty():
            return None
        else:
            toReturn = self.tail.payload
            if self.head == self.tail:
                self.head = self.tail = None
            else:
                tmp = self.head
                while tmp.next != self.tail:
                    tmp = tmp.next 
                tmp.next = None
                self.tail = tmp
            return toReturn

    def sort(self):
        outer = self.head
        while outer!= None:
            swapped = True
            while swapped == True:
                swapped = False
                inner = self.head
                while inner.next != None:
                    if inner.payload > inner.next.payload:
                        tmp = inner.payload
                        inner.payload = inner.next.payload
                        inner.next.payload = tmp
                        swapped = True
                    inner = inner.next
            outer = outer.next

    def deleteOnIndex(self, index):
        size = self.size()
        if index < 0 or index > size or self.isEmpty():
            return 
        else:
            if index == 0:
                self.deleteFromHead()
            elif index == size - 1:
                self.deleteFromTail()
            else:
                prev = None
                current = self.head
                counter = 0
                while counter < index:
                    prev = current
                    current = current.next
                    counter += 1
                prev.next = current.next

    def deleteNodesWithValue(self, value):
        prev = None
        tmp = self.head
        while tmp != None:
            if tmp.payload == value:
                if tmp == self.head:
                    self.deleteFromHead()
                elif tmp == self.tail:
                    self.deleteFromTail()
                else:
                    prev.next = tmp.next
                    tmp = prev
            else:
                prev = tmp
            
            tmp = tmp.next

    def insertAfter(self, listPayload, newPayload):
        tmp = self.head
        while tmp != None:
            if tmp.payload == listPayload:
                if tmp == self.tail:
                    self.addToTail(newPayload)
                else:
                    newNode = Node(newPayload, tmp.next)
                    tmp.next = newNode
                tmp = tmp.next
            tmp = tmp.next

    def insertBefore(self, listPayload, newPayload):
        prev = None
        tmp = self.head
        while tmp != None:
            if tmp.payload == listPayload:
                if tmp == self.head:
                    self.addToHead(newPayload)
                else:
                    newNode = Node(newPayload, tmp)
                    prev.next = newNode
            prev = tmp
            tmp = tmp.next


